File: 组合.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        ans, mid = [], []
        def dfs(i):
            if len(mid) == k:
                ans.append(mid.copy())
                return
            
            if i > k - len(mid):
                dfs(i - 1)
            mid.append(i)
            dfs(i - 1)
            mid.pop()
        dfs(n)
        return ans

File: 组合.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> ans;
        vector<int> mid;

        function<void(int)> dfs = [&](int i) -> void {
            if ((int)mid.size() == k) {
                ans.emplace_back(mid);
                return;
            }
            for (int j = i; j >= k - mid.size(); --j) {
                mid.emplace_back(j);
                dfs(j - 1);
                mid.pop_back();
            }
        };
        
        dfs(n);
        return ans;
    }
};
